Now that we got the document loaded & parsed , the next step is to calculate an "embedding" of that text. 
An embedding in it's simplest form is a multi dimensional mathematical vector that calculates the similarity of a piece of text. Topics and content that are similar have vectors that are close together. 

Embeddings enhance traditional search engines that use keywords or synonyms. Both have their place, but in this example we'll use embeddings as a way to calculate the proximity of texts.

Why similarity you might ask ? Well RAG allows us to provide and LLM more context about a subject when we ask it a question. This context is done by looking up documents that have similarity and giving that information. So in this section we're laying the groundwork for that end-goal.

In [2]:
# Install the necessary packages - tiktoken is used for calculating tokens
%pip install langchain langchain-openai


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In the following example we have a bunch of texts and calculate embeddings with them.
There are many different flavors of embeddings. Some work better with long texts, some better with multi lingual texts. Often llm providers will also provide an API to calculate embeddings through their service.
They are not per se the best for your use case.

In [3]:
# https://python.langchain.com/docs/modules/data_connection/text_embedding/

from langchain_openai import OpenAIEmbeddings

embeddings_model = OpenAIEmbeddings()

embeddings = embeddings_model.embed_documents(
    [
        "Hi there!",
        "Oh, hello!",
        "What's your name?",
        "My friends call me World",
        "Hello World!"
    ]
)
print(f"Embeddings calculated: {len(embeddings)}")
print(f"Vectorsize of first embedding :{len(embeddings[0])}")


/Users/patrick.debois/Library/Caches/pypoetry/virtualenvs/learning-llms-and-genai-for-dev-sec-ops-Sv6v5t2W-py3.11/lib/python3.11/site-packages/langchain_core/_api/deprecation.py:117: LangChainDeprecationWarning: The class `langchain_community.embeddings.openai.OpenAIEmbeddings` was deprecated in langchain-community 0.1.0 and will be removed in 0.2.0. An updated version of the class exists in the langchain-openai package and should be used instead. To use it run `pip install -U langchain-openai` and import as `from langchain_openai import OpenAIEmbeddings`.
  warn_deprecated(


Embeddings calculated: 5
Vectorsize of first embedding :1536


Next to calculating the embeddings from documents, we could also calculate that of a question.
We assume that through that we can find the documents that are best to provide the answer.
In this case the vector size is 1536.

In [3]:
# calculate the embedding of the query
embedded_query_results = embeddings_model.embed_query("What was the name mentioned in the conversation?")
print(f"Vectorsize of query text embedding :{len(embedded_query_results)}")

Vectorsize of query text embedding :1536
